### ISSUE TITLE
Bad glyphs in PDF when subsetting Noto fonts

### ISSUE THREAD

**Issue by box8** [galkahana/PDF-Writer#150]
Fonts (OpenType) taken from [https://www.google.com/get/noto/help/cjk/](https://www.google.com/get/noto/help/cjk/)

test.cpp:
```
#include "AbstractContentContext.h"
#include "PageContentContext.h"
#include "PDFPage.h"
#include "PDFWriter.h"

int main(int argc, char **argv)
{
	PDFWriter pdfWriter;
	pdfWriter.StartPDF("test.pdf", ePDFVersionMax);

	PDFUsedFont* font1 = pdfWriter.GetFontForFile("NotoSerifCJKjp-Regular.otf");
	PDFUsedFont* font2 = pdfWriter.GetFontForFile("NotoSerifCJKjp-Bold.otf");
	PDFUsedFont* font3 = pdfWriter.GetFontForFile("NotoSansCJKjp-Regular.otf");

	AbstractContentContext::TextOptions textOptions1(font1, 20, AbstractContentContext::eGray, 0);
	AbstractContentContext::TextOptions textOptions2(font2, 20, AbstractContentContext::eGray, 0);
	AbstractContentContext::TextOptions textOptions3(font3, 20, AbstractContentContext::eGray, 0);

	PDFPage* page = new PDFPage();
	page->SetMediaBox(PDFRectangle(0, 0, 595, 842));
	PageContentContext* contentContext = pdfWriter.StartPageContentContext(page);

	const char* test = "広告掲載";
	contentContext->WriteText(50, 750, test, textOptions1);
	contentContext->WriteText(50, 700, test, textOptions2);
	contentContext->WriteText(50, 650, test, textOptions3);

	pdfWriter.EndPageContentContext(contentContext);
	pdfWriter.WritePage(page);
	delete page;

	pdfWriter.EndPDF();
	return 0;
}
```

result: [test.pdf](https://github.com/galkahana/PDF-Writer/files/2730777/test.pdf)

check: [http://pdf-analyser.edpsciences.org/result/386239c9](http://pdf-analyser.edpsciences.org/result/386239c9)

**Comment by box8** [comment]
pdf-analyser.edpsciences.org data is expired, here's copy:

Name | Family | Sample Content | Embedded
AAAAAD+NotoSerifCJKjp-Regular | Noto Serif CJK JP | 広告掲載 | Yes
AAAAAC+NotoSerifCJKjp-Bold | Noto Serif CJK JP | 広告掲載 | Yes
AAAAAB+NotoSansCJKjp-Regular | Noto Sans CJK JP | 広告掲載 | Yes

characters are present in PDF file, glyphs are blank (or distorted in some PDF browsers).

**Comment by box8** [comment]
Some glyph sequences even cause segfault under Linux.
For example, [掤](https://unicode-table.com/en/63A4) + [接](https://unicode-table.com/en/63A5)
```
auto font = pdfWriter.GetFontForFile("NotoSerifCJKjp-Regular.otf");
pcc->BT();
pcc->Tf(font, 12);
pcc->Tm(1, 0, 0, 1, 50, 750);
pcc->Tj("\xE6\x8E\xA4\xE6\x8E\xA5"); // 掤接
pcc->ET();
```

**Comment by darcy-shen** [comment]
> Some glyphs even cause segfault in Linux.

This should be fixed now by #240 . But the exported PDF is still corrupted. FYI @galkahana

**Comment by darcy-shen** [comment]
> But the exported PDF is still corrupted.

The exported PDF is ok. It is corrupted because I didn't provide `NotoSerifCJKjp-Bold.otf`. `pdfWriter.GetFontForFile("NotoSerifCJKjp-Bold.otf");` should print warnings when the file does not exist.

**Comment by darcy-shen** [comment]
Confirmed, this bug is fixed by #240 too.
